rfcs/017 - Cluster Data Storage RFC.md

```diff
@@ -0,0 +1,221 @@
+
+The datastore node API should work similarly but at a higher level. The node determines based on LogRendezvousHash which node should own the namespace+key pair in question, then it fires that request to the *appropriate* node, which is the node in question under normal circumstances but if the internal bookkeeping of node routing has an intermediary node listed, it goes to that node instead. When handling a request for the data, the node receiving the request hashes the namespace+key pair and determines if it is the correct node for the job or if it should forward that request to another node, which will make this process seamless.
+
+This exact functionality can be automatically used by the `Anycloud` app by just requesting something like `/datastore/has/<namespace>/<key>` to the local instance and being sure that it will forward along the request to the correct node, or the local instance happens to actually own that data and handles it itself.
+
+Providing the cluster secret string into the Docker container as an environment variable along with the 4142 port is the other piece that makes it possible, and these requests can be put into a simple library for the language the Anycloud app is written in to cover up these implementation details and make it fit in better with the language in question.
+
+But those endpoints make clear something else that's necessary: the `HandlerMemory` structs that represent the value of a particular key need to be able to be serialized and deserialized, and in a way that they can be done so not just in Rust, but any language we want to fully support in Anycloud. (That's not entirely correct, we could create a set of secondary endpoints just for Anycloud that only accepts strings, or rather `Array<int8>` arrays where strings are a special case of that, but that may limit/impede future efforts at RPC between the Alan and Anycloud for offloading CPU-intensive work to Alan, and we still need serialization/deserialization of HandlerMemory for Alan itself.)
+
+Since all nodes in a cluster will be running the same version of Alan, which means they all have the same HandlerMemory, [protobuf](https://github.com/protocolbuffers/protobuf) is the best solution here for speed, size, and compatibility concerns, as it officially supports many languages and there is also a [Rust implementation](https://docs.rs/protobuf/2.22.1/protobuf/) that can be used. There are other protocols that are more widespread (JSON), faster (CapnProto), and smaller (Serde), but Protobuf hits a sweet spot where it's decent at all three (and it isn't an Apache project monstrosity, will veto any push for something out of apache.org).
+
+### Global Mutable State that Actually Works
```

Comment:
The distributed store compute should be something we should take into account for the compute or might be significant for the performance of the node and our "autoscaling"?

Reply:
I'm not sure I understand the question here?

Reply:
I think @aguillenv means how the different failure scenarios for cluster coordination affect how we autoscale?

Reply:
Yes, my comment it is non blocking but just to understand if is something that could affect significantly the load the system even when our autoscaler would manage it.

Reply:
Discussed offline, but for future reference, the load that this routing adds to the cluster would be monitored by the existing stats monitoring in the AVM, since that is at the server level, not process level, so the autoscaler should still kick in if the increased load from rerouting messages becomes a big deal.

Further, the actual amount of compute needed for the rerouting is very small and unlikely to be more than a rounding error, so it shouldn't ever be the trigger of the autoscaler, anyways.